src/main/java/org/apache/sysml/hops/BinaryOp.java

```diff
@@ -818,10 +819,24 @@ protected double computeIntermediateMemEstimate( long dim1, long dim2, long nnz
 			ret = getInput().get(0).getMemEstimate() * 3; 
 		}
 		else if ( op == OpOp2.SOLVE ) {
-			// x=solve(A,b) relies on QR decomposition of A, which is done using Apache commons-math
-			// matrix of size same as the first input
-			double interOutput = OptimizerUtils.estimateSizeExactSparsity(getInput().get(0).getDim1(), getInput().get(0).getDim2(), 1.0); 
-			return interOutput;
+			if (DMLScript.USE_ACCELERATOR) {
+				// Solve on the GPU takes an awful lot of intermediate space
+				// First the inputs are converted from row-major to column major
+				// Then a workspace and a temporary output (workSize, tauSize) are needed
+				long m = getInput().get(0).getDim1();
+				long n = getInput().get(0).getDim2();
+				long tauSize = OptimizerUtils.estimateSize(m, 1);
+				long workSize = OptimizerUtils.estimateSize(m, n);
+				long AtmpSize = OptimizerUtils.estimateSize(m, n);
+				long BtmpSize = OptimizerUtils.estimateSize(n, 1);
```

Comment:
Hm, are all these intermediates required in our JVM process? If not, we need to think about separating CP and GPU memory requirements - this is probably useful anyway to make the estimates tighter but it might be more involved as it would change the conditions for selecting GPU instructions.